File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        def bst(lo, hi):
            if lo > hi:
                return
            mid = int((hi + lo) / 2)
            node = TreeNode(nums[mid])
            node.left = bst(lo, mid-1)
            node.right = bst(mid+1, hi)
            return node
        return bst(0, len(nums)-1)

File: convert-sorted-array-to-binary-search-tree/convert-sorted-array-to-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        return generateBST(0, nums.length-1, nums);
    }
    
    public TreeNode generateBST(int lo, int hi, int[] nums) {
        if (lo > hi) {
            return null;
        }
        int mid = (lo + hi) / 2;
        var node = new TreeNode(nums[mid]);
        node.left = generateBST(lo, mid-1, nums);
        node.right = generateBST(mid+1, hi, nums);
        return node;
    }
}